v2/internal/identity/credential_provider.go

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package identity
+
+import (
+	"context"
+	"reflect"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	"github.com/Azure/azure-service-operator/v2/internal/config"
+	"github.com/Azure/azure-service-operator/v2/internal/reconcilers"
+	"github.com/Azure/azure-service-operator/v2/internal/util/kubeclient"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/core"
+)
+
+const (
+	// #nosec
+	globalCredentialSecretName = "aso-controller-settings"
+	// #nosec
+	NamespacedSecretName = "aso-credential"
+)
+
+// Credential describes a credential used to connect to Azure
+type Credential struct {
+	TokenCredential azcore.TokenCredential
+	CredentialFrom  types.NamespacedName
+	SubscriptionID  string
+
+	// secretData contains the secret
+	secretData map[string][]byte
+}
+
+func (c *Credential) SecretsEqual(other *Credential) bool {
+	return reflect.DeepEqual(c.secretData, other.secretData)
+}
+
+func NewDefaultCredential(tokenCred azcore.TokenCredential, namespace string, subscriptionID string) *Credential {
+	return &Credential{
+		TokenCredential: tokenCred,
+		SubscriptionID:  subscriptionID,
+		CredentialFrom:  types.NamespacedName{Namespace: namespace, Name: globalCredentialSecretName},
+	}
+}
+
+type CredentialProvider interface {
+	GetCredential(ctx context.Context, obj genruntime.MetaObject) (*Credential, error)
+}
+
+type credentialProvider struct {
+	globalCredential *Credential
+	kubeClient       kubeclient.Client
+}
+
+func NewCredentialProvider(
+	globalCredential *Credential,
+	kubeClient kubeclient.Client,
+) CredentialProvider {
+	return &credentialProvider{
+		kubeClient:       kubeClient,
+		globalCredential: globalCredential,
+	}
+}
+
+// GetCredential returns the credential used to manage the obj. It performs credential lookup according to the following
+// hierarchy:
+// 1. Per-Resource credential specified as an annotation (serviceoperator.azure.com/credential-from) on the obj
+// 2. Per-Namespace credential provided at the namespace scope ("aso-credential" secret in the obj's namespace)
+// 3. Global credential for the operator.
```

Comment:
Minor tweaking of wording for clarity

```suggestion
// GetCredential returns the credential used to manage the obj. It performs credential lookup according to the 
// following precedence:
// 1. Per-Resource credential specified as an annotation (`serviceoperator.azure.com/credential-from`) directly on the resource
// 2. Per-Namespace credential provided at the namespace scope (a secret called `aso-credential` in the resource's namespace)
// 3. Global credential for the operator.
```

I'm assuming annotations can't be inherited, so they have to be directly on the resource itself - is that right?

Reply:
Updated the docs here according to your suggestions.